File: Computer Science/Programming Language/Week04/SyntaxAnalyzer.java
import java.io.IOException;
import java.io.InputStreamReader;

public class rd_paresr {
  static char LA = 0;
  static InputStreamReader isr;

  static char nexttoken() throws IOException {
    return (char) isr.read();
  }

  static void match(char token) throws IOException {
    if (LA == token) LA = nexttoken();
    else error(token);
  }

  static void error(char c) {
    System.out.println("Syntax Error : '" + c + "'\n");
    System.exit(-1);
  }

  static void K() throws IOException {
    if ((int)LA == -1) return; // EOF
    else {
      S(); mathc('\n'); K();
    }
  }

  static void S() throws IOException {
    if(LA == '(') {
      match('(')
    }else ;
  }

  public static void main(String[] args) {
    isr = new InputStreamReader(System.in);
    LA = nexttoken();
    K();
  }

}

File: Computer Science/Programming Language/Week04/SyntaxAnalyzer.go
// Q. 영상을 참고해서 구문분석기를 Java 또는 Go로 구현하시오.

package main

import "fmt"

var lookahead int = 0

func match(c int){
   if c == lookahead {
      fmt.Scanf("%c", &lookahead)
   } else{
      fmt.Printf("Syntax error : ")
      fmt.Printf("%c\n", lookahead)
   }
}

func S(){
   if lookahead == '(' {
      match('(')
      S()
      match(')')
      S()
   }
}

func K(){
   if lookahead != 0 {
      S()
      match('\n')
      K()
   }
}

func main(){
   fmt.Scanf("%c", &lookahead)
   K();
}
